[IVDescriptors] Identify min/max recurrences in single pass.

fhahn opened the pull request:
This patch ties to extend the approach from
https://github.com/llvm/llvm-project/pull/141431 to all min/max recurrence kinds.

This patch adds a new getMinMaxRecurrence that identifies all min/max recurrences in a single pass. It starts at the backedge value of a phi and tries to identify the kind of the min/max recurrences.

It then walks from the backedge value to its operands recursively until it reaches out-of-loop values or the phi.

Then users of both the backedge value and all instructions in the chain from backedge value to phi are checked.

This consolidates all logic to identify min/max recurrences to a single function, and avoids the need to try to identify each min/max reduction kind individually.

fhahn committed:
[IVDescriptors] Identify min/max recurrences in single pass.

This patch ties to extend the approach from
https://github.com/llvm/llvm-project/pull/141431 to all min/max
recurrence kinds.

This patch adds a new getMinMaxRecurrence that identifies all min/max
recurrences in a single pass. It starts at the backedge value of a phi
and tries to identify the kind of the min/max recurrences.

It then walks from the backedge value to its operands recursively until
it reaches out-of-loop values or the phi.

Then users of both the backedge value and all instructions in the chain
from backedge value to phi are checked.

This consolidates all logic to identify min/max recurrences to a single
function, and avoids the need to try to identify each min/max reduction
kind individually.

fhahn reviewed (commented):
ping :)

alexey-bataev commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,173 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+  for (Use &U : RdxNext->uses()) {
+    auto *User = cast<Instruction>(U.getUser());
+    if (!TheLoop->contains(User->getParent())) {
+      if (++IncOut > 1)
```
```
if (IncOut > 0)
  return ...;
++IncOut
```

fhahn committed:
!fixup address comments, thanks

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,173 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+  return {collectMinMaxFMF(FPOp)};
+}
+
+static std::optional<RecurrenceDescriptor>
```
We need to return a `RecurrenceDescriptor`, but we don't need to use std::optional; we can check the recurrence kind is none, done thanks

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,174 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+              hasRequiredFastMathFlags(cast<FPMathOperator>(Cur), RK, FuncFMF))
+        FMF &= *CurFMF;
+      else
+        return {};
```
Could write with early-return?

artagnon reviewed (commented):
This is a great improvement. However, I still need time to validate the fresh logic.

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,174 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+      RK = RecurKind::FMinNum;
+    else
+      return std::nullopt;
+  }
```
Done, it is slightly more verbose, but perhaps a bit clearer.

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -412,18 +568,14 @@ bool RecurrenceDescriptor::AddReductionVar(
-    if (isIntMinMaxRecurrenceKind(Kind) && (isa<ICmpInst>(Cur) || IsASelect))
-      ++NumCmpSelectPatternInst;
-    if (isFPMinMaxRecurrenceKind(Kind) && (isa<FCmpInst>(Cur) || IsASelect))
-      ++NumCmpSelectPatternInst;
```
Yes, `NumCmpSelectPatternInst` can be removed once the new logic supports AnyOf as well.

IMO there are multiple advantages of the new structure:
* we can detect all min/max recurrences in a single pass, no need to try to detect the reduction pattern repeatedly for each kind
* more centralized logic, checking everyting that is needed for min/max in a single place, rather than intermingling with the other reduction kinds with differnet, specialized requirements.

fhahn committed:
!fixup adjust code, thanks

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -1035,24 +1101,13 @@ bool RecurrenceDescriptor::isReductionPHI(PHINode *Phi, Loop *TheLoop,
+  if (RD.getRecurrenceKind() != RecurKind::None) {
+    assert(
+        RecurrenceDescriptor::isMinMaxRecurrenceKind(RD.getRecurrenceKind()) &&
+        "must return a min/max recurrence kind");
```
```suggestion
        "Expected min/max recurrence kind");
```

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,184 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+          !SE->isLoopInvariant(Ptr, TheLoop) ||
+          (IntermediateStore &&
+           SE->getSCEV(IntermediateStore->getPointerOperand()) != Ptr))
+        return {};
```
This approximately corresponds to LVLegality::isInvariantAddressOfReduction, right? Perhaps add a suitable comment to this block?

artagnon replied:
I think you missed this comment?

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,184 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+      continue;
+    }
+    Value *A, *B;
+    RecurKind CurRK = GetMinMaxRK(Cur, A, B);
```
Where are A and B used? Why does GetMinMaxRK need to bind them?

artagnon reviewed (commented):
Okay, the logic looks sound, as far as I can tell.

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,184 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+  // Identify min/max recurrences by walking the def-use chains upwards,
+  // starting at RdxNext.
+  SmallVector<Value *> WorkList({RdxNext});
+  SmallPtrSet<Value *, 8> Chain({Phi});
```
AFAICT the iteration order doesn't matter below, as we need to validate all entries in the chain, and bail out if any one does not satisfy the requirement.

fhahn committed:
!fixup add message, refactor constructor

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -214,6 +214,184 @@ static bool checkOrderedReduction(RecurKind Kind, Instruction *ExactFPMathInst,
+      if (IncOut > 0)
+        return {};
+      IncOut++;
+    } else if (auto *SI = dyn_cast<StoreInst>(User)) {
```
make_filter_range for the case where TheLoop->contains() the User->getParent()? And then a range difference to get the rest?

fhahn committed:
!fixup remove left-over SmallPtrSet, addcomment

artagnon reviewed (approved):
LGTM, thanks!

fhahn committed:
[LV] Add additional tests for reductions with intermediate stores.

Adds missing test coverage for reductions with intermediate stores,
including partial reductions and additional intermediate stores.

fhahn committed:
!fixup handle all intermediate store cases.

fhahn committed:
!fixup simplify after merge.

fhahn committed:
Merge remote-tracking branch 'origin/main' into perf/ivdesc-steps

fhahn committed:
Merge remote-tracking branch 'origin/main' into perf/ivdesc-steps

alexfh commented:
Hi @fhahn , this commit causes an assertion failure on some real code: https://gcc.godbolt.org/z/fMe8ooav7
